### ISSUE TITLE
Support for Pytorch-Lightning multi-GPU

### ISSUE THREAD

**Issue by DomInvivo** [dmlc/dgl#3053]
## 🚀 Feature
Add multi-gpu support when models are trained using Pytorch-Lightning.

## Motivation
Pytorch-lightning is one of the most popular packages for training models in pytorch. It is simple, full-featured, and allows to handle distributed training easily. However, it currently crashes when training DGL models with a multi-gpu setting since the batches do not get converted appropriately. This is related to the objectives of DGL 0.7 to the roadmap of supporting pytorch-lightning #2888 .

## Additional context

I tried training a model on a single machine with multi-GPUs, so the training is not distributed, but rather the batches are split evenly and spread across GPUs for parallel computation.

There are 2 problems that I noticed:

- The `device` of the `DGLHeteroGraph` objects stays on the CPU and is not sent to the GPU automatically, as is the case with the labels. At each iteration, the `DataLoader` loads data from the RAM, and pytorch-lightning is supposed to send them to the GPU, but it does not happen for some reason.
- Manually pushing the device to the right GPU does not help either, since the batches are not split accordingly. For example, if I have a batch size of 128 on 2 GPUs, it should be split into 2 batches of size 64. Here, the batches are not split and each GPU receives the full batch of 128 elements.

**Comment by BarclayII** [comment]
Were you training a graph classification model or a node classification model?

For the former, currently DGL only supports DistributedDataParallel, not DataParallel.  So I guess you will need to try GraphDataLoader (with `use_ddp` option).

For the latter, currently DGL does not support multi-GPU training with PyTorch Lightning on node classification because DGL's node classification training requires forking the trainer processes from the main process to conserve memory, and as far as I know PyTorch Lightning does not support forking.

**Comment by hadim** [comment]
Thanks @BarclayII. 

> Were you training a graph classification model or a node classification model?

We use a graph classification model.


Indeed we got the error using `dp`. We also tried `ddp` but we were running into another error (we didn't keep track of the trace and we also weren't using `GraphDataLoader`). Will do more experiments and keep you in touch!
